components/api_version/src/lib.rs

```diff
@@ -88,6 +91,39 @@ macro_rules! match_template_api_version {
      }}
 }
 
+/// Dispatch an expression with type `kvproto::kvrpcpb::ApiVersion` to corresponding concrete type of `APIVersion`
+///
+/// For example, the following code
+///
+/// ```ignore
+/// let encoded_key = dispatch_api_version(api_version, API::encode_raw_key(key));
+/// ```
+///
+/// generates
+///
+/// ```ignore
+/// let encoded_key = match api_version {
+///     ApiVersion::V1 => APIV1::encode_raw_key(key),
+///     ApiVersion::V1ttl => APIV1TTL::encode_raw_key(key),
+///     ApiVersion::V2 => APIV2::encode_raw_key(key),
+/// };
+/// ```
+#[macro_export]
+macro_rules! dispatch_api_version {
```

Comment:
This seems overkill as it just a match.

Reply:
Without the macro, codes would be verbose, as all the match arms are almost the same. E.g., [`Simulator::run_node`](https://github.com/tikv/tikv/pull/12324/files#diff-8b71d304aa79611acfb1c46ecd474e6ae860f8fe140d7ad780cff49a3f47e32dR563) will need to be like this:

```rust
impl Simulator for ServerCluster {
    fn run_node(
        &mut self,
        node_id: u64,
        cfg: Config,
        engines: Engines<RocksEngine, RocksEngine>,
        store_meta: Arc<Mutex<StoreMeta>>,
        key_manager: Option<Arc<DataKeyManager>>,
        router: RaftRouter<RocksEngine, RocksEngine>,
        system: RaftBatchSystem<RocksEngine, RocksEngine>,
    ) -> ServerResult<u64> {
        match cfg.storage.api_version() {
            ApiVersion::V1 => {
              self.run_node_impl::<api_version::APIV1>(
                node_id,
                cfg,
                engines,
                store_meta,
                key_manager,
                router,
                system,
              )
           },
            ApiVersion::V1ttl => {
              self.run_node_impl::<api_versioni::APIV1TTL>(
                node_id,
                cfg,
                engines,
                store_meta,
                key_manager,
                router,
                system,
              )
           },
            ApiVersion::V2 => {
              self.run_node_impl::<api_version::APIV2>(
                node_id,
                cfg,
                engines,
                store_meta,
                key_manager,
                router,
                system,
              )
           },
        )
    }
```

And it would be worse if we have more api version enumerated values in the future.

Reply:
Since there aren't many places to dispatch API version, I suggest using the plain `match_template_api_version`

Reply:
En... what are the pros to keep sync with other `match_template_xxx` ?

There are 12 calls to `match_template_api_version` or `dispatch_api_version` (and there would be more in the future). And `dispatch_api_version` seems to be a little more clear.